### DISCUSSION TITLE
Removing an Pandoc Node/Inline Element

### DISCUSSION THREAD

**Discussion by simonvincenthk** [boisgera/pandoc#63] [Q&A]
Hi, 

I am writing a function to remove the nested links generated when reading an automatically-generated TOC with page numbers from a `*.docx` document:

Markdown Representation
```
[1 First-Level Heading [1](#first-level-heading)](#first-level-heading)
```

AST Representation
```
Link(('', [], []), [Str('1'), Space(), Str('First-Level'), Space(), Str('Heading'), Space(), Link(('', [], []), [Str('1')], ('#first-level-heading', ''))], ('#first-level-heading', ''))
```

Analogously to the [De-emphasize example](https://boisgera.github.io/pandoc/examples/#de-emphasize), I store all instances of nested links (`Link` nodes that occur inside other `Link` nodes) inside a `locations` list and then iterate through `locations` separately, in reverse order, to perform an action on each of it elements. 

Instead of removing the `Inline` contents from the node, I am interested in deleting the node all together:

```
for elt, holder, index in reversed(locations):
    assert isinstance(elt, pandoc.types.Link)
    # delete elt from doc
```

Each nested link is, however, both an instance of `Link` and an element in its parent `Link`'s `Inline` attribute. Accordingly, I'm having trouble removing it. I've tried `del elt`, `elt = None`, and `doc.remove(elt)`.

Is there a method or alternative way to remove an entire node from a `Pandoc`?

Help is appreciated!

Best regards, Simon

**Comment by boisgera** [comment]
Hi @simonvincenthk!

Once you have located where the nodes you want to remove are located -- what is the the node `holder` that contains them and in which position `index`, the classic `del holder[index]` works. The reverse iteration in the deletion phase is necessary not to mess up with the index of the nodes that are yet to be deleted.

There is no simple high-level API to remove nodes ATM  (a higher-level API was "in the works" but I had very little time to work on it in the previous years so I wouldn't hold my breath ...)

Anyway, something along the lines of the following work should work:

```python
import pandoc
from pandoc.types import Link, Str, Space

doc = Link(('', [], []), [Str('1'), Space(), Str('First-Level'), Space(), Str('Heading'), Space(), Link(('', [], []), [Str('1')], ('#first-level-heading', ''))], ('#first-level-heading', ''))

print(doc)
print(pandoc.write(doc, format="markdown")) 

inner_links_location = []
for elt, path in pandoc.iter(doc, path=True):
    if isinstance(elt, Link) and not elt is doc:
        holder, index = path[-1]
        inner_links_location.append([holder, index])

for holder, index in reversed(inner_links_location):
    del holder[index]

print(doc)
print(pandoc.write(doc, format="markdown")) 
```

Best regards,

Sébastien
